store/src/java/com/zimbra/cs/mailbox/Mailbox.java

```diff
@@ -10422,4 +10453,103 @@ public void noOp() throws ServiceException {
         // do nothing
     }
 
+    private SearchHistory getSearchHistory() throws ServiceException {
+        return SearchHistory.getFactory().getSearchHistory(getAccount());
+    }
+
+    public void addToSearchHistory(OperationContext octxt, String searchString) throws ServiceException {
+        boolean success = false;
+        try {
+            beginTransaction("addToSearchHistory", octxt);
+            SearchHistory searchHistory = getSearchHistory();
+            if (!searchHistory.contains(searchString)) {
+                int searchId = getNextSearchId(ID_AUTO_INCREMENT);
+                ZimbraLog.search.info("creating new search history entry '%s' with ID %d", searchString, searchId);
+                searchHistory.registerSearch(searchId, searchString);
+            }
+            ZimbraLog.search.info("adding '%s' to search history", searchString);
+            searchHistory.logSearch(searchString);
+            success = true;
+        } finally {
+            endTransaction(success);
+        }
+    }
+
+    public List<String> getSearchHistory(OperationContext octxt, SearchHistoryParams params) throws ServiceException {
+        boolean success = false;
+        try {
+            beginReadTransaction("getSearchHistory", octxt);
```

Comment:
why do we need read and write to search history be inside a transaction?

Reply:
I thought it was necessary for consistency (the searches table has an FK constraint on mailbox_id) and for redologs. I suppose reading doesn't have to be in a read transaction... but I worried about adding DB code that references mailboxes outside of the existing transaction mechanism.